apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt

```diff
@@ -95,14 +99,83 @@ fun ApolloClient.Builder.store(store: ApolloStore, writeToCacheAsynchronously: B
       .writeToCacheAsynchronously(writeToCacheAsynchronously)
 }
 
+enum class WatchErrorHandling {
+  THROW_CACHE_ERRORS,
+  THROW_NETWORK_ERRORS,
+  THROW_CACHE_AND_NETWORK_ERRORS,
+  IGNORE_ERRORS
+}
+
 /***
  * Gets the result from the network, then observes the cache for any changes.
  * Overriding the [FetchPolicy] will change how the result is first queried.
+ * Exception are ignored by default, this can be changed by setting [fetchErrorHandling] for the first fetch and [refetchErrorHandling]
+ * for subsequent fetches.
+ */
+@JvmOverloads
+fun <D : Query.Data> ApolloCall<D>.watch(
+    fetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+    refetchErrorHandling: WatchErrorHandling = WatchErrorHandling.IGNORE_ERRORS,
+): Flow<ApolloResponse<D>> {
+  var data: D? = null
+  return toFlow()
+      .catch {
+        maybeThrow(it, fetchErrorHandling)
+      }.onEach {
+        data = it.data
+      }.onCompletion {
+        emitAll(
+            watch(data)
+                .catch {
+                  maybeThrow(it, refetchErrorHandling)
+                }
+        )
+      }
+}
+
+/**
+ * Observes the cache for the given data. Unlike [watch], no initial request is executed on the network.
+ * The refetch policy set by [refetchPolicy] will be used.
+ * Cache and network exceptions are propagated.
  */
-fun <D : Query.Data> ApolloCall<D>.watch(): Flow<ApolloResponse<D>> {
-  return copy().addExecutionContext(WatchContext(true)).toFlow()
+fun <D : Query.Data> ApolloCall<D>.watch(data: D?): Flow<ApolloResponse<D>> {
+  return copy().addExecutionContext(WatchContext(data)).toFlow()
 }
 
+private fun maybeThrow(throwable: Throwable, errorHandling: WatchErrorHandling) {
+  // Only potentially swallow Apollo exceptions, the rest must be surfaced
+  if (throwable !is ApolloException) {
```

Comment:
Is that required? `Flow.catch` [rethrows the cancellation exceptions](https://github.com/Kotlin/kotlinx.coroutines/blob/ae0c842d9f5d8e4718d54960325c260dc551426e/kotlinx-coroutines-core/common/src/flow/operators/Errors.kt#L169) so we should only get our own exceptions there?

Reply:
I was thinking other "unexpected" exceptions (OOM?) should not be swallowed here? But maybe I'm overthinking it :)

Reply:
Good point about OOM 👍 . Can you add a doc to the enum?